File: atcoder_training/arc161/A_Make_M_intro_select.rs
//                A - Make M
// ----------------------------------------
// 問題
// https://atcoder.jp/contests/arc161/tasks/arc161_a
// ----------------------------------------

// attributes
#![allow(unused_imports)]
#![allow(unused_variables)]
#![allow(non_snake_case)]
#![allow(dead_code)]
#![allow(unused_macros)]

// imports
use itertools::Itertools;
use proconio::{
    fastout, input,
    marker::{Bytes, Chars, Usize1},
};

macro_rules! debug {
    ( $($val:expr),* $(,)* ) => {{
        #[cfg(debug_assertions)]
        eprintln!( concat!($(stringify!($val), " = {:?}, "),*), $($val),* );
    }};
}

pub mod intro_select {
    use itertools::Itertools;

    /// ## partition
    /// arrの要素をpivot以下とpivotより大きいものに分割
    pub fn partition<T: Ord + std::fmt::Debug>(arr: &mut [T], pivot: &T) -> Option<usize> {
        // pivotの位置
        let mut pivot_idx = 0;
        for (i, val) in arr.iter().enumerate() {
            if val == pivot {
                pivot_idx = i;
            }
        }
        if &arr[pivot_idx] != pivot {
            return None;
        }
        // 一番右のインデックス
        let r = arr.len() - 1;
        // pivotを一番右側に移動
        arr.swap(pivot_idx, r);
        // 移動していく
        let mut l = 0;
        for i in 0..r {
            if &arr[i] <= pivot {
                arr.swap(l, i);
                l += 1;
            }
        }
        // 末尾と交換
        arr.swap(l, r);
        Some(l)
    }

    /// ## select
    /// 配列arrのk番目の値を取り出す：O(N)
    pub fn select<T: Ord + Clone + std::fmt::Debug>(arr: &mut [T], k: usize) -> &T {
        if arr.len() == 1 {
            return &arr[0];
        }

        // 5つずつの要素にわけ、中央値をとる
        let size = (arr.len() + 4) / 5;
        let mut medians = {
            let mut medians = vec![];

            for i in 0..size {
                let begin = 5 * i;
                let end = arr.len().min(begin + 5);
                let med = &arr[begin .. end]
                    .iter()
                    .sorted()
                    .nth((end - begin - 1) / 2)
                    .unwrap();
                medians.push(med.clone().clone());
            }
            medians
        };

        // 中央値の中央値
        let pvt = select(&mut medians, (size - 1) / 2);

        // ピボットの位置
        let pvt_idx = partition(arr, pvt).unwrap();

        if pvt_idx < k {
            select(&mut arr[pvt_idx + 1..], k - pvt_idx - 1)
        } else if pvt_idx > k {
            select(&mut arr[..pvt_idx], k)
        } else {
            &arr[k]
        }
    }
}

use intro_select::*;

// main
fn main() {
    input! {
        N: usize,
        mut A: [usize; N],
    }

    if N == 1 {
        println!("Yes");
        return;
    }

    let x = *select(&mut A, N / 2 + 1);
    
    let cnt = A.iter().filter(|v| v == &&x).count();

    debug!(x, cnt);

    if cnt > N / 2 {
        println!("No");
    } else {
        println!("Yes");
    }
}

File: contests/arc161/A_Make_M_intro_select.py
#                A - Make M
# ----------------------------------------
# 問題
# https://atcoder.jp/contests/arc161/tasks/arc161_a
# ----------------------------------------

def partition(arr, pvt):
    p_idx = None
    for i, val in enumerate(arr):
        if val == pvt:
            p_idx = i
    if p_idx is None:
        return
    
    # pivotを末尾に
    arr[p_idx], arr[-1] = arr[-1], arr[p_idx]
    p_idx = -1

    # 順に交換
    for i, val in enumerate(arr[:-1]):
        if val <= pvt:
            p_idx += 1
            arr[p_idx], arr[i] = arr[i], arr[p_idx]
    
    # pivotを戻す
    arr[p_idx + 1], arr[-1] = arr[-1], arr[p_idx + 1]

    return p_idx + 1


def select(arr, k):
    """
    配列arrのk番目の要素を取得
    """

    if len(arr) == 1:
        return arr[0]

    # ピボット
    pvt = pivot(arr)

    # ピボットのある位置
    pvt_idx = partition(arr, pvt)

    # ピボットの位置で場合分け
    if pvt_idx < k:
        return select(arr[pvt_idx + 1:], k - pvt_idx - 1)
    elif pvt_idx > k:
        return select(arr[:pvt_idx], k)
    else:
        return arr[k]


def pivot(arr):
    """
    arrを5個ずつに分割した配列の中央値の中央値を求める
    """

    if len(arr) == 1:
        return arr[0]

    size = (len(arr) + 4) // 5
    medians = [0] * size

    for i in range(size):
        sub = arr[5 * i: 5 * (i + 1)]

        # 5この要素の中央値
        med = sorted(sub)[(len(sub) - 1) // 2]

        medians[i] = med
    
    # 中央値の中央値を検索
    return select(medians, (size - 1) // 2)


def main():
    N = int(input())
    A = list(map(int, input().split()))

    if N == 1:
        print("Yes")
        return

    x = select(A, N // 2 + 1)
    cnt = A.count(x)

    if cnt > N / 2:
        print("No")
    else:
        print("Yes")


main()
